### PR TITLE
Add gRPC service wrappers

### PR DESCRIPTION
**Pull request by mrfelton**
This implements a lot of the code that we use in Zap Desktop with regards to managing the lnd gRPC connections.

It supports on-the-fly lnd version detection and dynamically establishing connections using the most relevant rpc proto files based on the version of lnd that is being connected to.

### CODE REVIEW

File: src/grpc.js [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+  async onAfterDisconnect() {
+    debug('Disconnected from lnd gRPC service')
+  }
+
+  /**
+   * Connect to and activate the wallet unlocker api.
+   */
+  async onBeforeActivateWalletUnlocker() {
+    // Set up a listener that connects to the lightning interface as soon as the wallet has been unlocked.
+    this.services.WalletUnlocker.on('unlocked', () => this.activateLightning())
```
Comment (korhaliv) [inline src/grpc.js]: do we intentionally swallow potential arguments here ? e.g can it be used like 
```this.services.WalletUnlocker.on('unlocked', this.activateLightning.bind(this))```

File: src/grpc.js [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+  async connect() {
+    debug(`Connecting to lnd gRPC service`)
+
+    // Verify that the host is valid.
+    const { host } = this.options
+    await validateHost(host)
+
+    // Probe the services to determine the wallet state.
+    const walletState = await this.determineWalletState()
```
Comment (korhaliv) [inline src/grpc.js]: `determineWalletState` throws in some cases. Where does that situation is supposed to be handled?

File: src/services/lightning.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    // Determine most relevant proto version based on the api info.
+    const [closestProtoVersion, latestProtoVersion] = await Promise.all([
+      getClosestProtoVersion(info.version),
+      getLatestProtoVersion(),
+    ])
+
+    // Reconnect using best matching rpc proto if needed.
+    if (closestProtoVersion !== latestProtoVersion) {
+      this.debug('Found better match. Reconnecting using rpc.proto version: %s', closestProtoVersion)
+      this.service.close()
```
Comment (korhaliv) [inline src/services/lightning.js]: is `close`  synchronous ?

File: src/service.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+      }
+      throw e
+    }
+  }
+
+  /**
+   * Add promisified helper methods for each method in the gRPC service.
+   * Inspiration from https://github.com/altangent/lnd-async
+   * @param  {Object} rpc [description]
+   * @return {[type]}     [description]
```
Comment (korhaliv) [inline src/service.js]: to me it looks like it returns `undefined`

File: test/lnd.initWallet.test.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import { spawn } from 'child_process'
+import test from 'tape-promise/tape'
+import sinon from 'sinon'
+import { join } from 'path'
+import rimraf from 'rimraf'
+import LndGrpc from '../src'
+import { spawnLnd, killLnd, grpcOptions, seed } from './helpers/lnd'
+import { waitForFile } from '../src/utils'
```
Comment (korhaliv) [inline test/lnd.initWallet.test.js]: unused

File: test/lnd.initWallet.test.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import { spawn } from 'child_process'
```
Comment (korhaliv) [inline test/lnd.initWallet.test.js]: unused

File: test/lnd.initWallet.test.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import test from 'tape-promise/tape'
+import sinon from 'sinon'
```
Comment (korhaliv) [inline test/lnd.initWallet.test.js]: unused

File: test/lnd.initWallet.test.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import { join } from 'path'
```
Comment (korhaliv) [inline test/lnd.initWallet.test.js]: unused

File: test/lnd.initWallet.test.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import rimraf from 'rimraf'
```
Comment (korhaliv) [inline test/lnd.initWallet.test.js]: unused

File: test/lnd.unlockWallet.test.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
+import { spawn } from 'child_process'
```
Comment (korhaliv) [inline test/lnd.unlockWallet.test.js]: unused

File: test/lnd.unlockWallet.test.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+import test from 'tape-promise/tape'
+import sinon from 'sinon'
```
Comment (korhaliv) [inline test/lnd.unlockWallet.test.js]: unused

File: test/lnd.unlockWallet.test.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+import { join } from 'path'
```
Comment (korhaliv) [inline test/lnd.unlockWallet.test.js]: unused

File: test/lnd.unlockWallet.test.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+import rimraf from 'rimraf'
+import LndGrpc from '../src'
+import { spawnLnd, killLnd, grpcOptions, seed } from './helpers/lnd'
+import { waitForFile } from '../src/utils'
```
Comment (korhaliv) [inline test/lnd.unlockWallet.test.js]: unused

File: test/proto.test.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
+'use strict'
+
+import path from 'path'
```
Comment (korhaliv) [inline test/proto.test.js]: unused

File: test/grpc.test.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
+import { spawn } from 'child_process'
```
Comment (korhaliv) [inline test/grpc.test.js]: unused

File: test/lnd.genSeed.test.js [hunk 16]
```diff
@@ -0,0 +1,38 @@
+import { spawn } from 'child_process'
```
Comment (korhaliv) [inline test/lnd.genSeed.test.js]: unused

File: test/lnd.genSeed.test.js [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+import test from 'tape-promise/tape'
+import sinon from 'sinon'
```
Comment (korhaliv) [inline test/lnd.genSeed.test.js]: unused

File: test/lnd.genSeed.test.js [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+import { join } from 'path'
```
Comment (korhaliv) [inline test/lnd.genSeed.test.js]: unused

File: test/lnd.genSeed.test.js [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+import rimraf from 'rimraf'
```
Comment (korhaliv) [inline test/lnd.genSeed.test.js]: unused

File: test/lnd.genSeed.test.js [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+import LndGrpc from '../src'
+import { spawnLnd, killLnd, grpcOptions } from './helpers/lnd'
+import { waitForFile } from '../src/utils'
```
Comment (korhaliv) [inline test/lnd.genSeed.test.js]: unused

File: src/grpc.js [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+  async onBeforeActivateLightning() {
+    const { Lightning } = this.services
+
+    // First connect to the Lightning service.
+    await Lightning.connect()
+
+    // Fetch the determined version.
+    const { version } = Lightning
+
+    // Get a list of all other available and supported servicves.
```
Comment (korhaliv) [inline src/grpc.js]: Comment typo `servicves`

File: src/service.js [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+
+    // Wait for the cert to exist (this can take some time immediately after starting lnd).
+    if (waitForCert) {
+      await waitForFile(cert, 20000)
+    }
+
+    // Create ssl credentials to use with the gRPC client.
+    let creds = await createSslCreds(cert)
+
+    // Add macaroon to crenentials if service requires macaroons.
```
Comment (korhaliv) [inline src/service.js]: `crenentials`

File: test/grpc.test.js [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+
+test('initialize', t => {
+  t.plan(14)
+  const grpc = new LndGrpc(grpcOptions)
+  t.equal(grpc.state, 'ready', 'should start in the ready state')
+  t.equal(grpc.can('activateWalletUnlocker'), true, 'can activateWalletUnlocker')
+  t.equal(grpc.can('activateLightning'), true, 'can activateLightning')
+  t.equal(grpc.can('disconnect'), false, 'can not disconnect')
+  t.equal(grpc.options, grpcOptions, 'should store constructor options on the options property')
+  t.true(grpc.services, 'should have a servives property')
```
Comment (korhaliv) [inline test/grpc.test.js]: `servives`

File: src/grpc.js (hunk 2 above)
Comment (mrfelton) [reply src/grpc.js]: Any error thrown by `determineWalletState` should bubble up to the caller. So in this case, `connect` should throw the error back to the end user.

File: src/services/lightning.js (hunk 3 above)
Comment (mrfelton) [reply src/services/lightning.js]: Yep, as far as I understand it

**Comment by mrfelton** [comment]
Have updated all of those @korhaliv . Also added support for stream methods in `wrapAsync`.

**Pull request merged**